### ISSUE TITLE
Possible issue using MFA with role

### ISSUE THREAD

**Issue by jimcroft** [sparkleformation/sfn#121]
This could be an error on my part but here's the issue I'm having.

All our engineers have low privilege IAM accounts (MFA not required) and can assume various elevated privilege roles as needed (those roles require MFA).

Using the aws-mfa wrapper script works ok for our base users. So I can do the following:

```
aws-mfa
eval $(aws-mfa)
bundle exec sfn list
```

... and get the stack list. When I try to do something with the elevated privilege role:

```
$ AWS_STS_ROLE_ARN=arn:aws:iam::XXXXXXXXXXXX:role/ElevatedUserRole bundle exec sfn destroy ATestStack
[ERROR]: Failed to create remote API connection. Please validate configuration!
ERROR: Miasma::Error::ApiError::RequestError: Forbidden - InvalidClientTokenId: The security token included in the request is invalid.
```

My .sfn file contains:

```
    aws_access_key_id ENV['AWS_ACCESS_KEY_ID']
    aws_secret_access_key ENV['AWS_SECRET_ACCESS_KEY']
    aws_sts_token ENV['AWS_SECURITY_TOKEN']
    aws_sts_role_arn ENV['AWS_STS_ROLE_ARN']
    aws_region ENV['AWS_REGION']
```

Any ideas? The security token should be valid as it's used in the first example when listing stacks. It just seems not to work when using those STS creds and then trying to assume a role.

**Comment by chrisroberts** [comment]
Want to make sure I'm understanding the setup correctly: 

To assume privileged role, MFA is required (provided here by aws-mfa). The security token provided by aws-mfa is then used with the user's restricted credentials to assume the privileged role.

Is that a correct summary?

**Comment by jimcroft** [comment]
Yes that's correct.

With the AWS CLI tools this setup works with the following config:

```
in ~/.aws/credentials

[work]
aws_access_key_id=AKXXXXXXXXXXXXXXXXXX
aws_secret_access_key=somethingverysecret
region=us-east-1


in ~/.aws/config

[profile workpoweruser]
region = us-east-1
output = json
role_arn = arn:aws:iam::XXXXXXXXXXXXXXX:role/DevOpsPowerUser
source_profile = work
mfa_serial = arn:aws:iam::XXXXXXXXXXXXXXX:mfa/my.iam.user.name
```

Where the work profile has the creds for my low priv IAM account that can assume the other roles as needed.

**Comment by chrisroberts** [comment]
Great, thanks. Wanted to make sure I had the flow correct. I have a pending miasma-aws release to get out today. Will try and include this functionality within the release.

**Comment by jimcroft** [comment]
Awesome! Thanks.

**Comment by chrisroberts** [comment]
Alright, this was a bit of a rabbit hole but it ended up being a great one to follow. The `develop` branch currently has all this support merged in prepping for a release if you would like to give it a test run. It includes the latest miasma-aws release which adds session token support. So, if you are using something like aws-mfa you just need to use this in your .sfn config:

```ruby
Configuration.new do
  credentials do
    aws_sts_session_token ENV['AWS_SESSION_TOKEN']
```

I've also added in a new callback to handle MFA interactions. This may not matter if aws-mfa is being used for the aws-cli or other tools as well, but it does remove the requirement of an outside tool for getting proper authentication setup:

```ruby
Configuration.new do
  callbacks 'aws_mfa'
```

It will create a file in the working directory to store credentials for re-use until expiry. Cheers!

**Comment by jimcroft** [comment]
Just testing out the aws_mfa callback now (was trying out the aws-mfa wrapper solely for sfn) . It's working a charm so far. :+1:

**Comment by jimcroft** [comment]
Not sure if related to aws_mfa or stack_policy callback (I've got both setup) but using the develop branch I get an ArgumentError when trying to destroy a stack which does succeed:

```
$ AWS_STS_ROLE_ARN=arn:aws:iam::XXXXXXXXXXXXXXX:role/DevOpsFullAdmin be sfn destroy TestStack
[WARN]: Destroying Stack: TestStack
[Sfn]: Destroy listed stack? (Y/N): Y
ERROR: ArgumentError: wrong number of arguments (1 for 0)
$ AWS_STS_ROLE_ARN=arn:aws:iam::XXXXXXXXXXXXXXX:role/DevOpsFullAdmin be sfn list
Name                 Created                   Updated                   Status               Template Description
TestStack         2016-02-18 15:08:28 UTC   2016-02-18 16:26:06 UTC   DELETE_IN_PROGRESS   Base VPC for services
$
```

**Comment by chrisroberts** [comment]
Yeah, had to do with the `after` callback getting adjusted and not including the glob for the params. I also ref'd a PR against this on CLI values knocking out the config values (for Hash type values). Updating the `bogo-cli` version in your bundle will fix that up.

**Comment by chrisroberts** [comment]
Fix and enhancements in 2.1.2. Thanks for the report!
